src/normalize/normalize.rs

```diff
@@ -270,6 +273,68 @@ pub fn normalize_user_addrs(addrs: &[Addr], pid: u32) -> Result<NormalizedUserAd
 }
 
 
+/// Normalize `addresses` belonging to a process.
+///
+/// Normalize all `addrs` in a given process. The `addrs` array has to
+/// be sorted in ascending order or an error will be returned.
+///
+/// Unknown addresses are not normalized. They are reported as
+/// [`Unknown`] meta entries in the returned [`NormalizedUserAddrs`]
+/// object. The cause of an address to be unknown (and, hence, not
+/// normalized), could have a few reasons, including, but not limited
+/// to:
+/// - user error (if a bogus address was provided)
+/// - they belonged to an ELF object that has been unmapped since the
+///   address was captured
+///
+/// The process' ID should be provided in `pid`. To normalize addresses of the
+/// calling processes, `0` can be provided as a sentinel for the current
+/// process' ID.
+///
+/// Normalized addresses are reported in the exact same order in which the
+/// non-normalized ones were provided.
+pub fn normalize_user_addrs_sorted(addrs: &[Addr], pid: u32) -> Result<NormalizedUserAddrs> {
+    normalize_user_addrs_sorted_impl(addrs.iter().copied(), pid)
+}
+
+
+/// Reorder elements of `array` based on index information in `indices`.
+fn reorder<T, U>(array: &mut [T], indices: Vec<(U, usize)>) {
+    debug_assert_eq!(array.len(), indices.len());
+
+    let mut indices = indices;
+    // Sort the entries in `array` based on the indexes in `indices`
+    // (second member).
+    for i in 0..array.len() {
+        while indices[i].1 != i {
+            let () = array.swap(i, indices[i].1);
+            let idx = indices[i].1;
+            let () = indices.swap(i, idx);
```

Comment:
why separate idx variable, you do array.swap() without extra variable, would this just work;

```
let () = indices.swap(i, indices[i].1);
```

?

Reply:
Yeah, Rust didn't like that:
```
error[E0502]: cannot borrow `indices` as immutable because it is also borrowed as mutable
   --> src/normalize/normalize.rs:179:38
    |
179 |             let () = indices.swap(i, indices[i].1);
    |                      ----------------^^^^^^^------
    |                      |       |       |
    |                      |       |       immutable borrow occurs here
    |                      |       mutable borrow later used by call
    |                      mutable borrow occurs here
    |
help: try adding a local storing this argument...
   --> src/normalize/normalize.rs:179:38
    |
179 |             let () = indices.swap(i, indices[i].1);
    |                                      ^^^^^^^^^^
help: ...and then using that local as the argument to this call
   --> src/normalize/normalize.rs:179:22
    |
179 |             let () = indices.swap(i, indices[i].1);
    |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

:-(

Reply:
oh, some invisible references are involved then. I assumed `indices[i].1` is a copy of integer. Whatever, it's fine :)

Reply:
> I assumed indices[i].1 is a copy of integer.

It's also my understanding. I honestly can't make sense of what exactly the complaint is. Arguments should be evaluated strictly before the call is made, so I fail to see how there can be any conflict.